apps/controller/main.cpp

```diff
@@ -0,0 +1,186 @@
+                  uint16_t analog_inputs_values_idx[NUM_ANALOG_IN], uint32_t analog_inputs_sums[NUM_ANALOG_IN]) {
+  for (int i = 0; i < NUM_ANALOG_IN; i++) {
+    uint16_t val      = analog_inputs[i].read();
+    uint16_t curr_idx = analog_inputs_values_idx[i];
+
+    analog_inputs_sums[i] -= analog_inputs_values[i][curr_idx];
+    analog_inputs_values[i][curr_idx] = val;
+    analog_inputs_sums[i] += analog_inputs_values[i][curr_idx];
+
+    curr_idx++;
+    if (curr_idx > ANALOG_NUM_SAMPLE_AVERAGE) curr_idx = 0;
+
+    analog_inputs_values_idx[i] = curr_idx;
+  }
+}
+
+bool interval_passed(uint32_t start_time, uint32_t current_time, uint16_t interval) {
+  if (current_time > start_time) return (current_time - start_time) > interval;
+  // else the timer has wrapped around
+  else
+    return (MAX_UINT32 - start_time + current_time) > interval;
+}
+
+void prepare_data(uint8_t* outbuff, uint16_t& digital_inputs_values, uint32_t analog_inputs_sums[NUM_ANALOG_IN]) {
+  memcpy(outbuff, &digital_inputs_values, NUM_BYTES_DIGITAL_IN);
+
+  for (int i = 0; i < NUM_ANALOG_IN; i++) {
+    /*
+            Before the moving average array for each input is filled at least once, some elements will be zero,
+            but it's probably ok to assume it will be filled by the time we calculate the average, so we'll just
+            divide by ANALOG_NUM_SAMPLE_AVERAGE here
+            But testing is still needed
+    */
+    uint16_t avg = analog_inputs_sums[i] / ANALOG_NUM_SAMPLE_AVERAGE;
+    memcpy(outbuff + NUM_BYTES_DIGITAL_IN + i * sizeof(avg), &avg, sizeof(avg));
+  }
+}
+
+void clean_data(uint16_t& digital_inputs_values, uint16_t& digital_inputs_curr_values,
+                uint16_t digital_inputs_counts[NUM_DIGITAL_IN],
+                uint16_t analog_inputs_values[NUM_ANALOG_IN][ANALOG_NUM_SAMPLE_AVERAGE],
+                uint16_t analog_inputs_values_idx[NUM_ANALOG_IN], uint32_t analog_inputs_sums[NUM_ANALOG_IN]) {
+  digital_inputs_values      = 0;
+  digital_inputs_curr_values = 0;
+
+  memset(digital_inputs_counts, 0, sizeof(digital_inputs_counts[0]) * NUM_DIGITAL_IN);
+
+  memset(analog_inputs_values, 0, sizeof(analog_inputs_values[0][0]) * NUM_ANALOG_IN * ANALOG_NUM_SAMPLE_AVERAGE);
+  memset(analog_inputs_values_idx, 0, sizeof(analog_inputs_values_idx[0]) * NUM_ANALOG_IN);
+  memset(analog_inputs_sums, 0, sizeof(analog_inputs_sums[0]) * NUM_ANALOG_IN);
+}
+
+void send_to_pc(uint16_t& digital_inputs_values, uint16_t& digital_inputs_curr_values,
+                uint16_t digital_inputs_counts[NUM_DIGITAL_IN],
+                uint16_t analog_inputs_values[NUM_ANALOG_IN][ANALOG_NUM_SAMPLE_AVERAGE],
+                uint16_t analog_inputs_values_idx[NUM_ANALOG_IN], uint32_t analog_inputs_sums[NUM_ANALOG_IN]) {
+  static uint8_t outbuff[NUM_BYTES_DIGITAL_IN + NUM_BYTES_ANALOG_IN] = {0};
+
+  prepare_data(outbuff, digital_inputs_values, analog_inputs_sums);
+
+  // sending data out
+  for (unsigned i = 0; i < sizeof(outbuff); i++) {
```

Comment:
"unsigned i", pretty sure there should be an "int" somewhere in there, or something similar

Reply:
unsigned defaults to int, it's fine